src/eval/reduction.rs

```diff
@@ -278,6 +279,39 @@ fn reduce_with_witness_inner<F: LurkField, C: Coprocessor<F>>(
                         }};
                     }
 
+                    // An array, for performance reasons
+                    let unops_map: SmallVec<[(Ptr<F>, _); 11]> = smallvec![
+                        (c.car.ptr(), Op1::Car),
+                        (c.cdr.ptr(), Op1::Cdr),
+                        (c.commit.ptr(), Op1::Commit),
+                        (c.num.ptr(), Op1::Num),
+                        (c.u64.ptr(), Op1::U64),
+                        (c.comm.ptr(), Op1::Comm),
+                        (c.char.ptr(), Op1::Char),
+                        (c.open.ptr(), Op1::Open),
+                        (c.secret.ptr(), Op1::Secret),
+                        (c.atom.ptr(), Op1::Atom),
+                        (c.emit.ptr(), Op1::Emit),
+                    ];
+
+                    // An array, for performance reasons
+                    let binops_map: SmallVec<[(Ptr<F>, _); 15]> = smallvec![
+                        (c.cons.ptr(), Op2::Cons),
+                        (c.strcons.ptr(), Op2::StrCons),
+                        (c.hide.ptr(), Op2::Hide),
+                        (c.sum.ptr(), Op2::Sum),
+                        (c.diff.ptr(), Op2::Diff),
+                        (c.product.ptr(), Op2::Product),
+                        (c.quotient.ptr(), Op2::Quotient),
+                        (c.modulo.ptr(), Op2::Modulo),
+                        (c.num_equal.ptr(), Op2::NumEqual),
+                        (c.equal.ptr(), Op2::Equal),
+                        (c.less.ptr(), Op2::Less),
+                        (c.greater.ptr(), Op2::Greater),
+                        (c.less_equal.ptr(), Op2::LessEqual),
+                        (c.greater_equal.ptr(), Op2::GreaterEqual),
+                    ];
```

Comment:
Since these lookups are called so many times, I wonder if a `HashMap` would cut off the (small) performance drop completely

Reply:
Those lookups are called exactly once per evaluation, and in fact they’re more of an iteration than a lookup. If you look at the individual commit messages you’ll see I initially implemented this with a HashMap, and have the opportunity to reproduce the bench results that brought me here.

Reply:
But thanks for your question @arthurpaulino, it really inspired me to have a second look at Smallvec, which was actually hampering performance here.